harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/localfilesystem/LocalFilesystemHarvester.java

```diff
@@ -123,25 +124,23 @@ private HarvestResult align(Path root) throws Exception {
 			}
 		}
 		result = visitor.getResult();
-		List<String> idsForHarvestingResult = visitor.getIdsForHarvestingResult();
+		List<Integer> idsForHarvestingResult = visitor.getIdsForHarvestingResult();
 		if (!params.nodelete) {
 			//
 			// delete locally existing metadata from the same source if they
 			// were
 			// not in this harvesting result
 			//
-            List<Metadata> existingMetadata = context.getBean(MetadataRepository.class).findAllByHarvestInfo_Uuid(params.getUuid());
-            for (Metadata existingId : existingMetadata) {
+		    List<Integer> existingMetadata = context.getBean(MetadataRepository.class).findAllIdsBy(MetadataSpecs.hasHarvesterUuid(params.getUuid()));
+            for (Integer existingId : existingMetadata) {
 
 				if (cancelMonitor.get()) {
 					return this.result;
 				}
-
-				String ex$ = String.valueOf(existingId.getId());
-				if (!idsForHarvestingResult.contains(ex$)) {
-					log.debug("  Removing: " + ex$);
-					dataMan.deleteMetadata(context, ex$);
-					result.locallyRemoved++;
+                if (!idsForHarvestingResult.contains(existingId)) {
```

Comment:
You might consider changing existingMetadata to a set.  I requires a 1 time cost to convert it to a set but checking lists for the existence of a set is a linear search each time so this is quite inefficient:

    List<Integer> harvesterMetadataIds = context.getBean(MetadataRepository.class).findAllIdsBy(MetadataSpecs.hasHarvesterUuid(params.getUuid()));
    Set<Integer> existingMetadata = Sets.newHashSet(harvesterMetadataIds);